### PR TITLE
add internal/compress/lz4 test

### PR DESCRIPTION
**Pull request by vankichi**
Signed-off-by: vankichi <kyukawa315@gmail.com>

### Description:

- Test
    - create test for `internal/compress/lz4`
    - add comment for each exportable var/function for `internal/compress/lz4`
- Refactor
    - change to use named return at `CompressVector()` because to get an `error` which will be cause at defer function.

### Environment:

- Go Version: 1.14.4
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.0

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [x] I have added tests and benchmarks to cover my changes.
- [x] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/compress/lz4.go [hunk 1]
```diff
@@ -121,10 +132,13 @@ type lz4Writer struct {
 	w   io.WriteCloser
 }
 
+// Write returns the number of bytes written from p (0 <= n <= len(p)).
+// If any errors occurs, it will return an error.
 func (l *lz4Writer) Write(p []byte) (n int, err error) {
 	return l.w.Write(p)
 }
 
+// Close closes dst and w.
```
Comment (kevindiu) [inline internal/compress/lz4.go:141]:
```suggestion
// Close closes the writer.
```

File: internal/compress/lz4.go [hunk 2]
```diff
@@ -108,10 +116,13 @@ type lz4Reader struct {
 	r   io.Reader
 }
 
+// Read returns the number of bytes for read p (0 <= n <= len(p)).
+// If any errors occurs, it will return an error.
 func (l *lz4Reader) Read(p []byte) (n int, err error) {
 	return l.r.Read(p)
 }
 
+// Close closes src and r.
```
Comment (kevindiu) [inline internal/compress/lz4.go:125]:
```suggestion
// Close closes the reader.
```

File: internal/compress/lz4.go [hunk 3]
```diff
@@ -108,10 +116,13 @@ type lz4Reader struct {
...
 func (l *lz4Reader) Close() (err error) {
 	return l.src.Close()
```
Comment (kevindiu) [inline internal/compress/lz4.go:127]: maybe we forget to close `l.r`?

File: internal/compress/lz4.go [hunk 4]
```diff
@@ -121,10 +132,13 @@ type lz4Writer struct {
...
+// Write returns the number of bytes written from p (0 <= n <= len(p)).
+// If any errors occurs, it will return an error.
 func (l *lz4Writer) Write(p []byte) (n int, err error) {
 	return l.w.Write(p)
 }
 
+// Close closes dst and w.
 func (l *lz4Writer) Close() (err error) {
 	err = l.w.Close()
 	if err != nil {
```
Comment (kevindiu) [inline internal/compress/lz4.go:144]: I am not sure should we return err here. If error occurs the `dst` will not be closed.

File: internal/compress/lz4_test.go [hunk 5]
```diff
@@ -91,31 +53,36 @@ func TestNewLZ4(t *testing.T) {
...
+						gob, _ = NewGob()
+						return
+					}(),
+					compressionLevel: 0,
+					lz4:              lz4.New(),
+				},
+			},
+		},
+		{
+			name: "returns (nil, error) when option apply fails",
```
Comment (kevindiu) [inline internal/compress/lz4_test.go:73]: I think we need `returns (Compressor, nil) when option apply success` test case

Comment (vankichi) [reply internal/compress/lz4_test.go:73]: the first test pattern covers your suggestion.
it has `defaultLZ4Opts` and apply it.

File: internal/compress/lz4.go (hunk 4 above)
Comment (vankichi) [reply internal/compress/lz4.go:144]: Both of `dst` and `w` returns error.

File: internal/compress/lz4.go (hunk 3 above)
Comment (vankichi) [reply internal/compress/lz4.go:127]: `r` is `io.Reader` not `io.ReadCloser`.
we do not have to care about it.

File: internal/compress/lz4_test.go (hunk 5 above)
Comment (kevindiu) [reply internal/compress/lz4_test.go:73]: But I think it misses the test case to set the custom option 🤔

File: internal/compress/lz4.go (hunk 4 above)
Comment (kevindiu) [reply internal/compress/lz4.go:144]: I mean in line 145, if `l.w.Close()` returns error, line 148 will not executed and I am not sure if it is correct implementation.

File: internal/compress/lz4_test.go (hunk 5 above)
Comment (vankichi) [reply internal/compress/lz4_test.go:73]: well...
OK, I'll try it.

File: internal/compress/lz4.go (hunk 4 above)
Comment (vankichi) [reply internal/compress/lz4.go:144]: ```return errors.Wrap(l.dst.Close(), err.Error())```
It seems no problem. 🤔 
error of `l.w.Close()`will be wrapped with error of `l.dst.Close()`.

File: internal/compress/lz4_test.go [hunk 6]
```diff
@@ -171,39 +139,162 @@ func Test_lz4Compressor_CompressVector(t *testing.T) {
...
+							WriteFunc: w.Write,
+							FlushFunc: func() error {
+								return nil
+							},
+							CloseFunc: func() error {
+								return nil
+							},
+						}
+					},
+				},
```
Comment (hlts2) [inline internal/compress/lz4_test.go:214]:
```suggestion
```

File: internal/compress/lz4_test.go [hunk 7]
```diff
@@ -171,39 +139,162 @@ func Test_lz4Compressor_CompressVector(t *testing.T) {
...
+				compressionLevel: 0,
+				lz4: &lz4.MockLZ4{
+					NewWriterLevelFunc: func(w io.Writer, level int) lz4.Writer {
+						return &lz4.MockWriter{
+							WriteFunc: func(p []byte) (int, error) {
+								return 0, errors.New("write err")
+							},
+							FlushFunc: func() error {
+								return nil
+							},
```
Comment (hlts2) [inline internal/compress/lz4_test.go:241]:
```suggestion
```

File: internal/compress/lz4_test.go [hunk 8]
```diff
@@ -171,39 +139,162 @@ func Test_lz4Compressor_CompressVector(t *testing.T) {
...
+								return errors.New("close err")
+							},
+						}
+					},
+				},
+			},
+			want: want{
+				want: nil,
+				err:  errors.New("close err"),
+			},
```
Comment (hlts2) [inline internal/compress/lz4_test.go:314]: Could you please add test case that occurs error when Write or Flush and Clouse returns error ?

File: internal/compress/lz4_test.go [hunk 9]
```diff
@@ -793,33 +877,42 @@ func Test_lz4Writer_Close(t *testing.T) {
...
+					},
+				},
+			},
+			want: want{
+				err: errors.New("close err"),
+			},
+		},
 	}
 
 	for _, test := range tests {
```
Comment (hlts2) [inline internal/compress/lz4_test.go:936]: Could you please add test code for E2E ? (E2E dose not use mock)
for example is this.
https://github.com/vdaas/vald/blob/dcc3a7c3faf1e045170e0300279044fd4f8db9fd/internal/compress/gzip_test.go#L303

**Review by hlts2 — CHANGES_REQUESTED** [review]

File: internal/compress/lz4_test.go (hunk 8 above)
Comment (vankichi) [reply internal/compress/lz4_test.go:314]: ??? I already created.
what did you mean?

File: internal/compress/lz4_test.go (hunk 9 above)
Comment (vankichi) [reply internal/compress/lz4_test.go:936]: I'm not sure about that.
when I run `gotests` command, this test will not be created.

File: internal/compress/lz4_test.go (hunk 8 above)
Comment (hlts2) [reply internal/compress/lz4_test.go:297]: sorry. :bow:
I could not find the following test case...

```go
{
			name: "returns (nil, error) when Close fails",
			args: args{
				vector: []float32{0.1, 0.2, 0.3},
			},
			fields: fields{
				gobc: &MockCompressor{
					CompressVectorFunc: func(vector []float32) (bytes []byte, err error) {
						return []byte("vdaas/vald"), nil
					},
				},
				compressionLevel: 0,
				lz4: &lz4.MockLZ4{
					NewWriterLevelFunc: func(w io.Writer, level int) lz4.Writer {
						return &lz4.MockWriter{
							WriteFunc: func(p []byte) (int, error) {
								return 0, errors.New("write err")
							},
							FlushFunc: func() error {
								return nil
							},
							CloseFunc: func() error {
								return errors.New("close err")
							},
						}
					},
				},
			},
			want: want{
				want: nil,
				err:  errors.Wrap(errors.New("write err"),errors.New("close err").Error())
			},
		},
```

but you have no need to fix it, as it has coverage.
I was making a mistake.:bow:

**Review by hlts2 — DISMISSED** [review]
LGTM

**Review by kevindiu — APPROVED** [review]
LGTM

**Pull request merged**